MNT: Update Lo L1c pointing set dimensions and attribute naming

greglucas opened the pull request:
Based upon the pset variables named here: https://lasp.colorado.edu/galaxy/spaces/IMAP/pages/281316700/Remaining+PSET+Work

Updates all of the dimensions to be (epoch, esa_energy_step, spin_angle, off_angle). One question I still have is whether we want to fill in the actual spin_angle values instead of an integer bin number. This is what I would want, but I am unclear what the standard is for other instruments and the files provided had integer bin numbers (even though the variables don't have the `_bin` suffix). Thoughts on which approach to use?

greglucas committed:
MNT: Update Lo L1c pointing set dimensions and attribute naming

greglucas committed:
Add hae_lon and hae_lat depends

greglucas committed:
Update ena maps test

greglucas committed:
ENH/FIX: Lo: Use the pointing set hae longitude/latitude

This updates the mapping code to use the HAE longitude/latitude
from the pointing set directly rather than creating a grid.

vmartinez-cu commented:
HIT uses actual angle values for it's zenith and azimuth angle coordinates. I'm not sure what other instruments are doing. If the angle values are known, it seems worth adding in

greglucas committed:
MNT: Change from bin numbers to actual degrees

Rather than storing the bin numbers, store the actual centers of
the bins instead

greglucas commented:
> HIT uses actual angle values for it's zenith and azimuth angle coordinates. I'm not sure what other instruments are doing. If the angle values are known, it seems worth adding in

Updated!

vmartinez-cu commented on imap_processing/cdf/config/imap_lo_l1c_variable_attrs.yaml:
```diff
@@ -80,100 +97,170 @@ pivot_angle:
 
+esa_mode:
+  <<: *default
+  DEPEND_0: esa_energy_step
```
Is `esa_energy_step` a time varying variable? I believe only epoch or another time variable can be assigned to DEPEND_0. You might have to use DEPEND_1 here.  [ISTP guidance](https://spdf.gsfc.nasa.gov/istp_guide/vattributes.html#DEPEND_0:~:text=DEPEND_0%20%2D%2D%2D%20required%20for%20time%2Dvarying%20variables)

sdhoyt replied:
It's not time varying and you're right, this should be DEPEND_1, not DEPEND_0

greglucas replied:
That is quite confusing to me even after reading it a few times. I think I agree with you, but that would mean that this has a DEPEND_1 without a DEPEND_0? 🤷

greglucas replied:
Done

tmplummer commented on imap_processing/ena_maps/ena_maps.py:
```diff
@@ -634,15 +634,10 @@ class LoPointingSet(PointingSet):
+        # 0.1 degree bin spacing in both spin angle (longitude) and off angle (latitude)
+        bin_spacing_deg = 0.1
+        longitude_bin_centers = np.arange(0 + bin_spacing_deg / 2, 360, bin_spacing_deg)
+        latitude_bin_centers = np.arange(-2 + bin_spacing_deg / 2, 2, bin_spacing_deg)
```
Can't these be grabbed from the PSET Dataset?
```suggestion
        longitude_bin_centers = self.data["spin_angle"].data
        latitude_bin_centers = self.data["off_angle"].data
```

greglucas replied:
I reworked this to use the hae_longitude/latitude variables now instead, grabbing those directly from the PSET. (brought in a previous commit I had into this PR)

vmartinez-cu commented on imap_processing/cdf/config/imap_lo_l1c_variable_attrs.yaml:
```diff
@@ -80,100 +97,170 @@ pivot_angle:
+  DEPEND_3: off_angle
   FIELDNAM: Double coincidence counts
   FORMAT: I12
   LABLAXIS: double coincidence counts
```
[LABLAXIS](https://spdf.gsfc.nasa.gov/istp_guide/vattributes.html#LABLAXIS:~:text=LABLAXIS%20%2D%2D%2D%20required%20if%20not%20using%20LABL_PTR_1) should be short character strings (6-10 characters), though I'm not sure what would happen if it's longer. It's hard to tell from the ISTP guidance what's a recommendation vs. a hard requirement.

vmartinez-cu replied:
HIT has some long values too, so maybe it's fine since Andriy didn't comment on it

greglucas replied:
Done

tmplummer commented on imap_processing/lo/l1c/lo_l1c.py:
```diff
@@ -68,15 +68,12 @@ def lo_l1c(sci_dependencies: dict, anc_dependencies: list) -> list[xr.Dataset]:
-            "longitude": np.arange(3600),
-            "latitude": np.arange(40),
+            "esa_energy_step": np.arange(1, 8),
+            "spin_angle": np.arange(3600) / 3600 + 0.05,
```
These don't look right. I prefer the way you implemented it above...
```suggestion
            "spin_angle": np.arange(0 + bin_spacing_deg / 2, 360, bin_spacing_deg),
            "off_angle": np.arange(-2 + bin_spacing_deg / 2, 2, bin_spacing_deg),
```

Or from your suggestion on Hi code, should you be using np.linspace. I forget the exact reason that you made that suggestion.

greglucas replied:
linspace is supposed to be more stable for floating point numbers according to the docs. I have upated this with globals to correspond with the Hi version.

tmplummer commented on imap_processing/lo/l1c/lo_l1c.py:
```diff
@@ -338,38 +333,54 @@ def create_datasets(
-            attrs=attr_mgr.get_variable_attributes("pointing_bins_label"),
+
+        spin_angle = xr.DataArray(
+            data=np.arange(3600) / 3600 + 0.05,
```
Here is another instance where this array is being generated. Perhaps put this in a module level global variable. The implementation here is not correct.

tmplummer commented on imap_processing/lo/l1c/lo_l1c.py:
```diff
@@ -338,38 +333,54 @@ def create_datasets(
         )
+
+        off_angle = xr.DataArray(
+            data=np.arange(40) / 40 - 1.95,
```
Likewise, put this into a module level global and fix implementation?

greglucas replied:
Yep, moved these up as globals, good suggestion!

vmartinez-cu commented on imap_processing/cdf/config/imap_lo_l1c_variable_attrs.yaml:
```diff
@@ -80,100 +97,170 @@ pivot_angle:
+  LABL_PTR_2: spin_angle_label
+  LABL_PTR_3: off_angle_label
+
+o_background_rates:
```
You can add DELTA_MINUS_VAR and DELTA_PLUS_VAR attribute fields and assign the appropriate uncertainty variables to these. This is optional according to [ISTP guidance](https://spdf.gsfc.nasa.gov/istp_guide/vattributes.html#DELTA:~:text=DELTA_PLUS_VAR%20and%20DELTA_MINUS_VAR%20%2D%2D%2D%20optional). I added these for HIT using total uncertainty values.

vmartinez-cu replied:
I forgot that this is L1, so it might not be worth the effort adding these in, but would be useful to include in L2

sdhoyt replied:
@vmartinez-cu They want this product to be archived, so do you think given that, the DELTA_VAR attributes should be added?

greglucas replied:
I added these to point at the uncert variable for both plus and minus per the recommendation.

tmplummer commented on imap_processing/cdf/config/imap_lo_l1c_variable_attrs.yaml:
```diff
@@ -7,69 +7,86 @@ default_attrs: &default
-  FIELDNAM: Pointing Bins
-  UNITS: ' '
+  VALIDMAX: 360
+  CATDESC: Spin angle in despun pointing frame
```
Might want to specify that these are bin centers.

tmplummer commented on imap_processing/cdf/config/imap_lo_l1c_variable_attrs.yaml:
```diff
@@ -7,69 +7,86 @@ default_attrs: &default
+off_angle:
+  VALIDMIN: -2
+  VALIDMAX: 2
+  CATDESC: Off axis angle in despun pointing frame
```
Specify that these are bin centers?

greglucas replied:
Done

tmplummer reviewed (commented):
I'd mostly like to see a common definition of the angular bin centers defined. For Hi I did it here: https://github.com/IMAP-Science-Operations-Center/imap_processing/blob/dev/imap_processing/hi/hi_l1c.py#L38

tmplummer commented:
@greglucas 
> Updates all of the dimensions to be (epoch, esa_energy_step, spin_angle, off_angle). One question I still have is whether we want to fill in the actual spin_angle values instead of an integer bin number. This is what I would want, but I am unclear what the standard is for other instruments and the files provided had integer bin numbers (even though the variables don't have the _bin suffix). Thoughts on which approach to use?

I would also want the angle bins. For Lo where the off_angle dimension does not cover the full latitude range, I would argue it is critical to specify angle.

I'm sure Paul will be amenable to me changing Hi to specify angle instead of bin number in the PSET product.

greglucas committed:
Update Lo attributes for ISTP compliance

greglucas committed:
MNT: Use global pset size definitions

greglucas committed:
MNT: Specifying bin centers for attributes

greglucas commented:
Thanks for the great reviews! I think I've addressed all comments now. I reworked the bin angles/centers code quite a bit, so would be good to review that and see if folks agree.

greglucas commented:
@sdhoyt It looks like codecov is failing because the `create_datasets` function is not called at all. Do you anticipate needing that or should I just remove it from here?

tmplummer reviewed (approved):
LGTM

sdhoyt commented:
> @sdhoyt It looks like codecov is failing because the `create_datasets` function is not called at all. Do you anticipate needing that or should I just remove it from here?

@greglucas That will be needed to set all the attributes or some of the code from that function will need to be pulled out and integrated into the other functions in the file. I don't think I've touched that function since I created an empty file for SIT-3, so it's out of date. I was leaving it there to use later, but the attributes weren't my priority so I stopped calling it rather than updating it as I made algorithm changes. Since the version history is there, it's probably fine to remove for now.

greglucas commented:
> @greglucas That will be needed to set all the attributes or some of the code from that function will need to be pulled out and integrated into the other functions in the file. I don't think I've touched that function since I created an empty file for SIT-3, so it's out of date. I was leaving it there to use later, but the attributes weren't my priority so I stopped calling it rather than updating it as I made algorithm changes. Since the version history is there, it's probably fine to remove for now.

OK, I'm fine merging over the codecov failures if you'll be using it in the future so I can just leave it for now would be my preference in that case.

Are you OK with me merging this now or do you want me to wait until yours is in so you don't have to deal with any conflicts? It looks like this is pretty separate from yours though, so merge conflicts wouldn't be too bad hopefully.

greglucas committed:
MNT: Use PSET Shape

greglucas committed:
Add in commented out attributes
